include/boost/gil/image_processing/threshold.hpp

```diff
@@ -334,6 +343,115 @@ void threshold_optimal
     }
 }
 
+namespace detail {
+
+template
+<
+    typename SourceChannelT,
+    typename ResultChannelT,
+    typename SrcView,
+    typename DstView,
+    typename Operator
+>
+void adaptive_impl
+(
+    SrcView const& src_view,
+    SrcView const& convolved_view,
+    DstView const& dst_view,
+    Operator const& threshold_op
+)
+{
+    //template argument validation
+    gil_function_requires<ImageViewConcept<SrcView>>();
+    gil_function_requires<MutableImageViewConcept<DstView>>();
+
+    static_assert(color_spaces_are_compatible
+    <
+        typename color_space_type<SrcView>::type,
+        typename color_space_type<DstView>::type
+    >::value, "Source and destination views must have pixels with the same color space");
+
+    //iterate over the image chaecking each pixel value for the threshold
+    for (std::ptrdiff_t y = 0; y < src_view.height(); y++)
+    {
+        typename SrcView::x_iterator src_it = src_view.row_begin(y);
+        typename SrcView::x_iterator convolved_it = convolved_view.row_begin(y);
+        typename DstView::x_iterator dst_it = dst_view.row_begin(y);
+
+        for (std::ptrdiff_t x = 0; x < src_view.width(); x++)
+        {
+            static_transform(src_it[x], convolved_it[x], dst_it[x], threshold_op);
+        }
+    }
+}
+} //namespace boost::gil::detail
+
+template <typename SrcView, typename DstView>
+void threshold_adaptive
+(
+    SrcView const& src_view,
+    DstView const& dst_view,
+    typename channel_type<DstView>::type max_value,
+    std::size_t kernel_size,
+    threshold_adaptive_method method = threshold_adaptive_method::mean,
+    threshold_direction direction = threshold_direction::regular
+)
+{
+    BOOST_ASSERT_MSG((kernel_size % 2 != 0), "Kernel size must be an odd number");
+
+    typedef typename channel_type<SrcView>::type source_channel_t;
+    typedef typename channel_type<DstView>::type result_channel_t;
+
+    if (method == threshold_adaptive_method::mean)
+    {
+        float *mean = new float[kernel_size];
+        std::fill_n(mean, kernel_size, 1.0f/kernel_size);
+
+        image<typename SrcView::value_type> temp_img(src_view.width(), src_view.height());
+        typename image<typename SrcView::value_type>::view_t temp_view = view(temp_img);
+        SrcView temp_conv(temp_view);
+        
+        kernel_1d<float> kernel(mean, kernel_size, kernel_size/2);
+        convolve_rows<pixel<float, typename SrcView::value_type::layout_t>>(
+            src_view, kernel, temp_view
+        );
+        convolve_cols<pixel<float, typename SrcView::value_type::layout_t>>(
```

Comment:
@miralshah365 Since #347, you should now be able to replace `convolve_rows` and `convolve_cols` with single:

```
convolve<pixel<float, typename SrcView::value_type::layout_t>>(src_view, kernel, temp_view);
```